src/DataflowType/Dataflow/Dataflow.php

```diff
@@ -67,16 +93,31 @@ public function process(): Result
                 $writer->prepare();
             }
 
-            foreach ($this->reader as $index => $item) {
-                try {
-                    $this->processItem($item);
-                } catch (\Throwable $e) {
-                    $exceptions[$index] = $e;
-                    $this->logException($e, (string) $index);
+            $deferred = new Deferred();
+            $resolved = false; //missing $deferred->isResolved() in version 2.5
+            $producer = new Producer(function (callable $emit) {
+                foreach ($this->reader as $index => $item) {
+                    yield new Delayed($this->emitInterval);
+                    yield $emit([$index, $item]);
+                }
+            });
+
+            Loop::repeat($this->loopInterval, function () use ($deferred, &$resolved, $producer, &$count, &$exceptions) {
```

Comment:
Shouldn't this Loop::repeat be explicitly cancelled at one point ?

Reply:
The Amp event loop `start` when you call Loop::run() and `stop` when you call Loop::stop(). This is managed at line 120 when calling `Amp\Promise\wait`

Loop::repeat does nothing until you start Loop::run.
When $deferred->resolve() is called,  then the promise act `Amp\Promise\wait` to stop Amp loop by calling Loop::stop.
When Amp stop the loop, there could be some ticks remaining, that why I introduced $resolved boolean flag. But for further version of Amp, this will be directly available in $deferred->isResolved()

Reply:
Sorry, my question was not precise enough.

According to the doc https://amphp.org/amp/event-loop/api#repeat, repeat loops must be explicitly cancelled to free their resources.

If there is another call to `Loop::run()` after a dataflow is processed (possibly in another dataflow processing), will this repeat watcher continue to run, doing nothing ?